server/src/main/java/org/apache/cassandra/sidecar/acl/authentication/MutualTlsAuthenticationHandler.java

```diff
@@ -88,10 +91,19 @@ public void authenticate(RoutingContext ctx, Handler<AsyncResult<User>> handler)
 
                         List<String> identities = extractIdentities(authN.result());
                         List<String> roles = extractCassandraRoles(identities);
-                        if (!roles.isEmpty())
+                        String roleIntended = ctx.request().getHeader(AUTH_ROLE);
+
+                        if (isNotEmpty(roleIntended) && !roles.contains(roleIntended))
                         {
-                            authN.result().attributes().put(CASSANDRA_ROLES_ATTRIBUTE_NAME, roles);
+                            String errMsg = String.format("None of the identities %s are authorized for role %s",
+                                                          identities, roleIntended);
+                            handler.handle(Future.failedFuture(wrapHttpException(UNAUTHORIZED, errMsg)));
+                            return;
                         }
+
+                        List<String> rolesToAdd = roleIntended != null && !roleIntended.isEmpty()
```

Comment:
NIT, since you are already using it
```suggestion
                        List<String> rolesToAdd = isNotEmpty(roleIntended)
```

Reply:
Updated it